### PR TITLE
update convert

### PR DESCRIPTION
**Pull request by mxyng**
the output of convert remains exactly the same

### CODE REVIEW

File: convert/convert.go [hunk 1]
```diff
@@ -279,51 +302,192 @@ func GetTensorName(n string) (string, error) {
...
+
+		bufSize := uint64(10240)
+		var finished bool
+		for {
+			data := make([]byte, min(bufSize, remaining))
+
+			b, err := io.ReadFull(f, data)
+			remaining -= uint64(b)
+
+			if err == io.EOF || remaining <= 0 {
```
Comment (bmizerany) [inline convert/convert.go:419]: errors.Is?

File: llm/gguf.go [hunk 2]
```diff
@@ -71,647 +72,179 @@ const (
...
+	return &gguf{
+		containerGGUF: container,
 		KV:            make(KV),
 	}
 }
 
-func (llm *GGUFModel) NumTensor() uint64 {
-	if llm.Version == 1 {
+func NewGGUFV3(bo binary.ByteOrder) *gguf {
+	return newGGUF(&containerGGUF{ByteOrder: bo, Version: 3})
```
Comment (BruceMacD) [inline llm/gguf.go:107]: this adds support for gguf v3 also?

File: llm/gguf.go [hunk 3]
```diff
@@ -38,21 +31,29 @@ type ContainerGGUF struct {
 	}
 }
 
-func (c *ContainerGGUF) Name() string {
+func (c *containerGGUF) Name() string {
```
Comment (jmorganca) [inline llm/gguf.go:34]: Can we move this to a `gguf` package in the future – would be better if this `llm` package was focused on the runner

Comment (mxyng) [reply llm/gguf.go:35]: I intended the `llm` package to encapsulate any large language model related code including the model (parsing, conversion, writing) in addition to running

File: convert/convert.go [hunk 4]
```diff
@@ -279,51 +302,192 @@ func GetTensorName(n string) (string, error) {
 	return "", fmt.Errorf("couldn't find a layer name for '%s'", n)
 }
 
+type safetensorWriterTo struct {
```
[the remark this answers is not in the document]
Comment (mxyng) [reply convert/convert.go]: sorry I should've mentioned. this is lifted almost verbatim from llm/gguf.go

**Review by BruceMacD — APPROVED** [review]

**Comment by mxyng** [comment]
superseded by #3250

**Pull request closed without merging**